crates/sui-graphql-rpc/src/types/move_package.rs

```diff
@@ -415,11 +460,53 @@ impl MovePackage {
         }
     }
 
+    /// Look-up the package by its Storage ID, as of a given checkpoint.
+    pub(crate) fn by_id_at(checkpoint_viewed_at: u64) -> PackageLookup {
+        PackageLookup::ById {
+            checkpoint_viewed_at,
+        }
+    }
+
+    /// Look-up a specific version of the package, identified by the storage ID of any version of
+    /// the package, and the desired version (the actual object loaded might be at a different
+    /// object ID).
+    pub(crate) fn by_version(version: u64, checkpoint_viewed_at: u64) -> PackageLookup {
+        PackageLookup::Versioned {
+            version,
+            checkpoint_viewed_at,
+        }
+    }
+
     pub(crate) async fn query(
         ctx: &Context<'_>,
         address: SuiAddress,
-        key: ObjectLookup,
+        key: PackageLookup,
     ) -> Result<Option<Self>, Error> {
+        let (address, key) = match key {
+            PackageLookup::ById {
+                checkpoint_viewed_at,
+            } => (address, Object::latest_at(checkpoint_viewed_at)),
```

Comment:
thought: are there specific fields we want from the objects table that lead us to using `Object::latest_at`, or can we fetch all the necessary data off the packages table?

Reply:
Originally, I tried to rely only on `packages`, but unfortunately it doesn't work because we can convert any `MovePackage` into an `Object` and that requires having all the fields of the `Object` around. It also means that we need to provide the same consistency guarantees for packages as we do for objects in case they get converted.

I originally did have a plan to use `packages` for this -- I think you only need to store the previous transaction digest to make it work (you can create an instance of `StoredHistoryObject` from an instance of `StoredPackage` that has all those fields), but the consistency requirements is what ruins things.  Instead, I'm usin `packages` analogously to `objects_version` here.

I also had a plan to re-use the package resolver's cache for loading packages in this way (or vice versa), but this realisation makes that impossible too.